### ISSUE TITLE
Configured "external" option not honored when using cache

### ISSUE THREAD

**Issue by olance** [rollup/rollup#3874]
### Expected Behavior / Situation
With two distinct configurations for building a CommonJS and a UMD bundle from the same source:
```js
const cjsBuild = {
    inputOptions: {
        input: 'src/index.js',
        external: ['./lib/utils'] // <-- Note declared external here
    },

    outputOptions: {
        format: 'cjs',
        name: 'Test',
        dir: './dist',
        entryFileNames: 'test.cjs.js'
    }
}
const umdBuild = {
    inputOptions: {
        input: 'src/index.js',
        external: [], // <-- Note empty external here
    },

    outputOptions: {
        format: 'umd',
        name: 'Test',
        dir: './dist',
        entryFileNames: 'test.umd.js'
    }
}
```

Using the JS API to build, with caching:

```js
let cache;

async function buildWithCache(config, useCache) {
    const bundle = await rollup({
        cache: useCache ? cache : false, // is ignored if falsy
        ...config
    });
    cache = bundle.cache; // store the cache object of the previous build
    return bundle;
}

buildWithCache(cjsBuild.inputOptions, true)
    .then(bundle => {
        return bundle.write(cjsBuild.outputOptions)
    })
    .then(() => buildWithCache(umdBuild.inputOptions, true)) // will use the cache of the previous build
    .then(bundle => {
        return bundle.write(umdBuild.outputOptions)
    })
```

Given the configuration for both build, I'd expect the CJS bundle to contain only my "main" code without the `lib/utils` imported code, and my UMD bundle to contain everything, since its config does not declare `./lib/utils` as external.

Similarly, if I were to reverse the build order, I'd expect the above description to remain true.

### Actual Behavior / Situation

What I see instead, is that the cache retains the external "nature" of any resolved/cached module, and uses that information as-is in subsequent builds when a cache is provided.

As a result:
- if I build the CJS bundle before the UMD one:
    - the CJS module does not include the code of my imported lib ✅
    - neither does the UMD module ❌
- if I build the UMD bundle before the CJS one:
    - the UMD module includes the code of my imported lib ✅
    - so does the CJS module ❌

I have created a simple repro repository to provide a full example: [olance/rollup-cache-external-bug](https://github.com/olance/rollup-cache-external-bug/tree/main)

### Modification Proposal

Would it be possible, even when using the cache, to reevaluate the external nature of a module based on the actual build configuration that's been provided?

**Comment by lukastaegert** [comment]
That is correct, resolution information is stored in the cache as file resolution is a very expensive process, and external detection is currently deeply nested into that process.

**Comment by olance** [comment]
@lukastaegert thanks for the quick reply!

So is the current situation the target behavior?
Or would you accept a PR to correct it to the expectations I described?

**Comment by lukastaegert** [comment]
Depends how you want to do that. Do you just want to honour the use-defined external option? That could be doable I guess, but not sure if in a clean way. If you want to respect files made external by plugins as well then you need to run the whole resolution, which is definitely something to avoid.

So a PR is welcome, but note that it might be more complicated than it appears. Also note that files are run through the external option twice, once before resolution and once after that.

**Comment by olance** [comment]
Understood 🙂 

I definitely don't know the internals of Rollup well enough (if at all ^^) to get that kind of insight, so thank you!
I'll try to have a look at it and see if I can pull up something :)
